File: scripts/sandbox_smoke.py
from __future__ import annotations

import asyncio
import json
import tempfile
from pathlib import Path

from agent_system.config import Settings
from agent_system.sandbox import DockerSandboxProvider, SandboxSpec


async def smoke() -> dict:
    with tempfile.TemporaryDirectory(prefix="agent-system-sandbox-") as raw:
        root = Path(raw)
        settings = Settings(
            environment="production",
            auto_create_schema=False,
            database_url="postgresql+asyncpg://unused:unused@localhost/unused",
            run_state_secret="sandbox-smoke-production-state-secret",
            sandbox_root=root,
        )
        provider = DockerSandboxProvider(settings)
        handle = await provider.create(
            SandboxSpec(
                run_id="sandbox-smoke",
                cpu_limit=0.5,
                memory_mb=128,
                disk_mb=64,
                timeout_seconds=20,
                network_enabled=False,
            )
        )
        try:
            identity = await provider.exec(handle, ["id", "-u"])
            workspace_write = await provider.exec(
                handle, ["sh", "-c", "printf ok > /workspace/result.txt"]
            )
            root_write = await provider.exec(
                handle, ["sh", "-c", "printf forbidden > /root-write-test"]
            )
            network = await provider.exec(
                handle,
                [
                    "python",
                    "-c",
                    "import socket; socket.create_connection(('1.1.1.1', 53), timeout=2)",
                ],
                timeout=5,
            )
            return {
                "container_id": handle.id,
                "non_root": identity.return_code == 0 and identity.stdout.strip() == "65534",
                "workspace_writable": workspace_write.return_code == 0,
                "root_filesystem_read_only": root_write.return_code != 0,
                "network_blocked": network.return_code != 0,
                "workspace_result": (root / "sandbox-smoke" / "result.txt").read_text(),
            }
        finally:
            await provider.destroy(handle)


if __name__ == "__main__":
    print(json.dumps(asyncio.run(smoke()), ensure_ascii=False, indent=2))

File: scripts/sandbox-smoke.ts
import { mkdtemp, rm } from "node:fs/promises";
import { tmpdir } from "node:os";
import { join } from "node:path";
import { getSettings } from "../src/config.js";
import { Container } from "../src/container.js";

const root = await mkdtemp(join(tmpdir(), "omoikane-sandbox-smoke-"));
const settings = getSettings({
  AGENT_DATABASE_URL: "pglite://:memory:",
  AGENT_ARTIFACT_ROOT: join(root, "artifacts"),
  AGENT_SKILL_ROOT: join(root, "skills"),
  AGENT_SANDBOX_ROOT: join(root, "sandboxes"),
  AGENT_CREDENTIAL_SECRET: "sandbox-test-secret",
  OMOIKANE_SANDBOX_PROVIDER: "docker",
});
const container = await Container.create(settings, { startWorker: false });
const handle = await container.sandbox.create({
  runId: "sandbox-smoke",
  networkEnabled: false,
  timeoutSeconds: 30,
  image: "node:22-alpine",
});
try {
  const skill = await container.skills.importBundle([
    {
      path: "SKILL.md",
      content_base64: Buffer.from(
        `---
name: Docker smoke Skill
slug: docker-smoke
omoikane:
  schema_version: 1
  workspace: required
  requires:
    tools: [sandbox_exec]
    commands: [node]
    network: false
  entrypoints:
    run:
      command: [node, .omoikane/skills/docker-smoke/scripts/run.mjs]
---

# Docker smoke

Run only through the explicitly bound Sandbox Tool.
`,
      ).toString("base64"),
    },
    {
      path: "input.txt",
      content_base64: Buffer.from("sandbox Skill input").toString("base64"),
    },
    {
      path: "scripts/run.mjs",
      content_base64: Buffer.from(
        "import fs from 'node:fs'; console.log(fs.readFileSync('.omoikane/skills/docker-smoke/input.txt','utf8')); console.log(process.getuid?.());\n",
      ).toString("base64"),
    },
  ]);
  const bindings = await container.skills.prepareForRun(
    [{ version_id: skill.version.id }],
    {
      toolNames: new Set(["sandbox_exec"]),
      sandboxConfig: { enabled: true, network_enabled: false },
      sandbox: handle,
      sandboxService: container.sandbox,
    },
  );
  const result = await container.sandbox.execute(handle, "node", [
    ".omoikane/skills/docker-smoke/scripts/run.mjs",
  ]);
  process.stdout.write(
    `${JSON.stringify(
      { provider: handle.provider, skill: bindings[0], ...result },
      null,
      2,
    )}\n`,
  );
  if (
    result.exit_code !== 0 ||
    !result.stdout.includes("sandbox Skill input") ||
    result.stdout.trim().endsWith("0") ||
    bindings[0]?.workspace !== ".omoikane/skills/docker-smoke"
  )
    process.exitCode = 1;
} finally {
  await container.sandbox.destroy(handle);
  await container.close();
  await rm(root, { recursive: true, force: true });
}
